### PR TITLE
[CPU] Refactor graph transformations interface

### PR DESCRIPTION
**Pull request by EgorDuplensky**
such as nodes / edges addition and removal

### Details:
 - use std algorithms where possible
 - remove duplication and unnecessary calls 

This is a first part of the planning changes.
The second part would be rethinking the DropNode() method usage

### CODE REVIEW

File: src/plugins/intel_cpu/src/graph_optimizer.cpp [hunk 1]
```diff
@@ -2536,16 +2485,7 @@ void GraphOptimizer::MergeTransposeAndReorder(Graph &graph) {
         auto childChildNode = childNode->getChildEdgeAt(0)->getChild();
 
         auto remEdge = parentNode->getParentEdgesAtPort(1)[0];
-        remEdge->drop();
-        auto& edges = graph.GetEdges();
-        for (auto it = edges.begin(); it != edges.end(); it++) {
-            if ((*it) == remEdge) {
-                edges.erase(it);
-                if (parentParentConstNode->getChildEdges().empty())
-                    parentParentConstNode->remove();
```
Comment (EgorDuplensky) [inline src/plugins/intel_cpu/src/graph_optimizer.cpp:2545]: @luo-cheng2021 @ceciliapeng2011 
Could you please take a look a this particular change.
I didn't get why we need to ->remove() the const input explicitly.
From my understanding edge->drop() is enough in this case.

File: src/plugins/intel_cpu/src/node.cpp [hunk 2]
```diff
@@ -211,43 +211,22 @@ void Node::addEdge(const EdgeWeakPtr& edge) {
...
-            break;
-        }
-    }
-    for (auto it = parentPtr->childEdges.begin(); it != parentPtr->childEdges.end(); it++) {
-        auto childEdge = (*it).lock();
-        if (childEdge && childEdge->getChild() == childPtr && childEdge->getParent() == parentPtr) {
-            parentPtr->childEdges.erase(it);
-            break;
+void Node::remove() {
+    // auto parent_edges = parentEdges;
```
Comment (EgorDuplensky) [inline src/plugins/intel_cpu/src/node.cpp:219]: remove comment

File: src/plugins/intel_cpu/src/graph_optimizer.cpp (hunk 1 above)
Comment (luo-cheng2021) [reply src/plugins/intel_cpu/src/graph_optimizer.cpp:2545]: Yes, it should be ok.

File: src/plugins/intel_cpu/src/graph.cpp [hunk 3]
```diff
@@ -1426,14 +1411,24 @@ void Graph::GetPerfData(std::vector<ov::ProfilingInfo>& perfMap) const {
     }
 }
 
+void Graph::CreateEdge(const NodePtr& parent,
+                       const NodePtr& child,
+                       int parentPort,
+                       int childPort) {
+    auto edge = std::make_shared<Edge>(parent, child, parentPort, childPort);
```
Comment (dmitry-gorokhov) [inline src/plugins/intel_cpu/src/graph.cpp:1418]: Sounds like it worth to add some validation for port values (at least debug only):
1. Non negative value
2. Child doesn't have parent edge at this port

File: src/plugins/intel_cpu/src/graph.cpp [hunk 4]
```diff
@@ -1426,14 +1411,24 @@ void Graph::GetPerfData(std::vector<ov::ProfilingInfo>& perfMap) const {
...
-            edge->drop();
-            graphEdges.erase(it);
-            return;
-        }
-    }
+    edge->drop();
+    graphEdges.erase(std::remove(graphEdges.begin(), graphEdges.end(), edge), graphEdges.end());
+}
+
+void Graph::AddNode(NodePtr node) {
```
Comment (dmitry-gorokhov) [inline src/plugins/intel_cpu/src/graph.cpp:1429]: Same comment: check on nullptr?

File: src/plugins/intel_cpu/src/graph_optimizer.cpp [hunk 5]
```diff
@@ -646,19 +638,15 @@ void GraphOptimizer::FuseConvolutionMatMulDeconvAndBias(Graph &graph) {
                     graph.InsertNode(biasNode, targetNode, cpuReshapeNode, inNum, outNum, false);
                     // Insert the Reshape const input node and edge into CPU graph.
                     const auto cpuReshapeConstInput = std::make_shared<node::Input>(reshapeConstInput, graph.getGraphContext());
-                    EdgePtr newReshapeConstEdge(new Edge(cpuReshapeConstInput, cpuReshapeNode, 0, 1));
-                    cpuReshapeNode->addEdge(newReshapeConstEdge);
-                    graphEdges.push_back(newReshapeConstEdge);
-                    graphNodes.push_back(cpuReshapeConstInput);
```
[the remark this answers is not in the document]
Comment (EgorDuplensky) [reply src/plugins/intel_cpu/src/graph_optimizer.cpp:652]: Good Idea, done.

File: src/plugins/intel_cpu/src/graph.cpp (hunk 3 above)
Comment (EgorDuplensky) [reply src/plugins/intel_cpu/src/graph.cpp:1421]: Done

File: src/plugins/intel_cpu/src/graph.cpp (hunk 4 above)
Comment (EgorDuplensky) [reply src/plugins/intel_cpu/src/graph.cpp:1433]: Done

**Review by dmitry-gorokhov — APPROVED** [review]

**Pull request merged**